sklearn/neighbors/_classification.py

```diff
@@ -211,9 +211,21 @@ def predict(self, X):
         y : ndarray of shape (n_queries,) or (n_queries, n_outputs)
             Class labels for each data sample.
         """
+        # Duplicated because of the check on self.effective_metric_'s value
+        # TODO: remove check_is_fitted duplication
+        check_is_fitted(self)
+
         X = self._validate_data(X, accept_sparse="csr", reset=False)
 
-        neigh_dist, neigh_ind = self.kneighbors(X)
+        if self.weights == "uniform" and self.effective_metric_ == "fast_euclidean":
+            # In that case, it is safe to use the fast alternative which
+            # does not use sqrt on distances as this can be costly.
+            self.effective_metric_ = "fast_sqeuclidean"
+            neigh_dist, neigh_ind = self.kneighbors(X)
+            self.effective_metric_ = "fast_euclidean"
+        else:
+            neigh_dist, neigh_ind = self.kneighbors(X)
```

Comment:
I don't like this hack. Would it make the code a lot more complex to do 
```
if self.weights == "uniform" and self.effective_metric_ == "fast_euclidean":
    self.effective_metric_ = "fast_sqeuclidean"
```
where it's defined and then check the metric and take the square root of the distances if we need to return the true distances ?

Reply:
To be honest, I don't like this hack too.

The idea here is to make use of `fast_sqeuclidean` for some calls (the ones which aren't returning) without changing the result on the induction made for the value of `self.effective_metric_` as the value has to be used for other calls. `fast_sqeuclidean` might help getting better performances, but I think that it might be marginal gain: still it's useful, but I think we should not favor small optimizations over clarity.

Such optimizations might also preferably be treated at a lower level (e.g. in `kneighbors` or even in `{ArgKmin, RadiusNeighborhood}.compute` if the metric is "fast_euclidean"` and `return_distance=False`).

Reply:
Ok maybe it's to painful to change the whole logic of effective_metric_. I agree however that it might be better to do it in PairwiseDistancesReduction

Reply:
Maybe introducing a private method`self._kneighbors(X, metric)` would help?

Reply:
f75f08e boils it down.